Split out CUDA RPC tests

lw opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #55778 Skip RPC's CPU-only tests on CircleCI GPU jobs
* **#55695 Split out CUDA RPC tests**

In order to be able to run CUDA tests on their own (e.g., to avoid running CPU tests on GPU machines).

Done by moving test methods to a separate class (and sometimes introducing a "common" base class for utils), and then providing new entry points inside a `cuda/` subdirectory.

Differential Revision: [D27618198](https://our.internmc.facebook.com/intern/diff/D27618198/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27618198/)!

lw committed:
Split out CUDA RPC tests

In order to be able to run CUDA tests on their own (e.g., to avoid running CPU tests on GPU machines).

Done by moving test methods to a separate class (and sometimes introducing a "common" base class for utils), and then providing new entry points inside a `cuda/` subdirectory.

Differential Revision: [D27618198](https://our.internmc.facebook.com/intern/diff/D27618198/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27618198/)!

[ghstack-poisoned]

mrshenli commented on .jenkins/pytorch/multigpu-test.sh:
```diff
@@ -22,5 +22,5 @@ time python test/run_test.py --verbose -i distributed/test_jit_c10d
 time python test/run_test.py --verbose -i distributed/test_c10d
 time python test/run_test.py --verbose -i distributed/test_c10d_spawn
-time python test/run_test.py --verbose -i distributed/rpc/test_tensorpipe_agent
+time python test/run_test.py --verbose -i distributed/rpc/cuda/test_tensorpipe_agent
```
I just realized that we are not running `process_group` tests on multi-gpu CI. Is this intentional?

lw replied:
Not sure, the TensorPipe tests were added in https://github.com/pytorch/pytorch/pull/49210 by @pritamdamania87. I think that TensorPipe has some "native" multi-GPU tests, whereas for ProcessGroup (which is CPU-only) it would just be the DDP part that is multi-GPU. Perhaps this was the reason not to add them here?

mrshenli replied:
Yep, it could be. And since we are retiring ProcessGroup soon, it should be fine to leave it out.

mrshenli reviewed (approved):
LGTM! Thanks for re-organizing tests.

lw committed:
Update on "Split out CUDA RPC tests"


In order to be able to run CUDA tests on their own (e.g., to avoid running CPU tests on GPU machines).

Done by moving test methods to a separate class (and sometimes introducing a "common" base class for utils), and then providing new entry points inside a `cuda/` subdirectory.

Differential Revision: [D27618198](https://our.internmc.facebook.com/intern/diff/D27618198/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27618198/)!

[ghstack-poisoned]
